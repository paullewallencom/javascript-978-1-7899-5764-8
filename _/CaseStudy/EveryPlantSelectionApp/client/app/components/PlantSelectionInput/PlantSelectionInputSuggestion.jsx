import React, {useState, useRef, useEffect} from 'react';

export default ({ plant, isSelected, onClick, onMouseEnter }) => {

  const ref = useRef();
  const [didSelectByMouseEnter, setDidSelectByMouseEnter] = useState(false);

  useEffect(() => {
    if (isSelected && !didSelectByMouseEnter) {
      ref.current.scrollIntoView(false);
    }
    if (!isSelected) {
      setDidSelectByMouseEnter(false);
    }
  }, [isSelected]);

  return (
    <li
      ref={ref}
      onMouseDown={onClick}
      className={
        'PlantSelectionInputSuggestion' + (isSelected ? ' isSelected' : '')
      }
      onMouseEnter={e => {
        setDidSelectByMouseEnter(true);
        onMouseEnter(e);
      }}
    >
      <span className="PlantSelectionInputSuggestion_genus_species">
        {plant.genus}
        {' '}
        {plant.species}
      </span>
      <span className="PlantSelectionInputSuggestion_family">
        {plant.family}
      </span>
    </li>
  );
};
