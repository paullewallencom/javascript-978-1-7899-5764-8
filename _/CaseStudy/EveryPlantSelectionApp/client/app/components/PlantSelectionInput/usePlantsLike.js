import {useState, useEffect} from 'react';
import debounce from 'debounce-promise';
import fetch  from 'node-fetch';

const makeRequest = debounce(query => {
  return (
      fetch(`http://localhost:3000/plants/${query}`)
        .then(response => response.json())
  );
}, 300, { leading: true });

export default (query, generateDataURL) => {

  const [loading, setLoading] = useState(false);
  const [plants, setPlants] = useState([]);

  useEffect(() => {

    if (!query) {
      setPlants([]);
      return;
    }

    setLoading(true);

    makeRequest(query).then(data => {
      setLoading(false);
      setPlants(data);
    });

  }, [query]);

  return {
    loading,
    plants
  }

};
