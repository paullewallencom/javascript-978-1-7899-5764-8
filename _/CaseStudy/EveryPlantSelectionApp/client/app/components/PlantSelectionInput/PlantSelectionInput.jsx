import React, {useState, useRef, useEffect} from 'react';
import PlantSelectionInputSuggestion from './PlantSelectionInputSuggestion.jsx';
import usePlantsLike from './usePlantsLike';

const MIN_QUERY_LENGTH = 3;

const KEYCODES = {
  ENTER: 13,
  ARROW_DOWN: 40,
  ARROW_UP: 38
};

export default ({ isInitiallityOpen, value, onSelect }) => {

  const inputRef = useRef();

  const [isOpen, setIsOpen] = useState(isInitiallityOpen || false);
  const [query, setQuery] = useState(value);

  const [selectedIndex, setSelectedIndex] = useState(1);
  const {loading, plants} = usePlantsLike(query);

  function onKeyUp(e) {
    switch (e.keyCode) {
      case KEYCODES.ENTER:
        if (selectedIndex > -1) {
          onSelect(plants[selectedIndex]);
        }
        break;
      case KEYCODES.ARROW_UP:
        setSelectedIndex(selectedIndex === 0 ? 0 : selectedIndex - 1);
        break;
      case KEYCODES.ARROW_DOWN:
        setSelectedIndex(
          selectedIndex === plants.length - 1 ? selectedIndex : selectedIndex + 1
        );
        break;
    }
  }

  useEffect(() => {
    setSelectedIndex(-1);
  }, [loading, plants]);

  return <div className="PlantSelectionInput">
    <input
      onFocus={() => {
        setQuery(inputRef.current.value);
        setIsOpen(true);
      }}
      onBlur={() => setIsOpen(false)}
      onChange={() => setQuery(inputRef.current.value)}
      ref={inputRef}
      autoComplete="off"
      aria-autocomplete="true"
      spellCheck="false"
      aria-expanded={String(isOpen)}
      role="combobox"
      onKeyUp={onKeyUp}
      defaultValue={value} />
    {
      isOpen &&
        <ol>
          {
            !plants.length && (!query || query.length < MIN_QUERY_LENGTH) &&
              <li className="PlantSelectionInput_notice">
                Please begin typing a plant name...
              </li>
          }
          {
            !plants.length && query && query.length >= MIN_QUERY_LENGTH && loading &&
              <li className="PlantSelectionInput_notice">
                Loading...
              </li>
          }
          {
            !plants.length && query && query.length >= MIN_QUERY_LENGTH && !loading &&
              <li className="PlantSelectionInput_notice">
                No plants with that name found...
              </li>
          }
          {
            plants.map(
              (plant, index) =>
              <PlantSelectionInputSuggestion
                key={plant.id}
                plant={plant}
                isSelected={selectedIndex === index}
                onMouseEnter={() => setSelectedIndex(index)}
                onClick={() => {
                  onSelect(plant);
                }}
              />
            )
          }
        </ol>
    }
  </div>;
};
