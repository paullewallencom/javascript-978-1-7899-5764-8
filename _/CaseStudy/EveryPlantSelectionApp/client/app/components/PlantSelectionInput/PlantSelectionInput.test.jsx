import React from 'react';
import renderer from 'react-test-renderer';
import PlantSelectionInput from './';

describe('PlantSelectionInput', () => {

  it('Should render deterministically to its snapshot', () => {
    expect(
      renderer
        .create(
          <PlantSelectionInput
            onSelect={() => {}}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  describe('With configured isInitiallyOpen & value properties', () => {
    it('Should render deterministically to its snapshot', () => {
      expect(
        renderer
          .create(
            <PlantSelectionInput
              isInitiallyOpen={true}
              value="Example..."
              onSelect={() => {}}
            />
          )
          .toJSON()
      ).toMatchSnapshot();
    });
  });

});
