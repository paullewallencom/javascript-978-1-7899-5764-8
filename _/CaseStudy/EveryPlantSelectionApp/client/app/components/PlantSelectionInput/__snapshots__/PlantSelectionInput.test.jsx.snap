// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantSelectionInput Should render deterministically to its snapshot 1`] = `
<div
  className="PlantSelectionInput"
>
  <input
    aria-autocomplete="true"
    aria-expanded="false"
    autoComplete="off"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    role="combobox"
    spellCheck="false"
  />
</div>
`;

exports[`PlantSelectionInput With configured isInitiallyOpen & value properties Should render deterministically to its snapshot 1`] = `
<div
  className="PlantSelectionInput"
>
  <input
    aria-autocomplete="true"
    aria-expanded="false"
    autoComplete="off"
    defaultValue="Example..."
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    role="combobox"
    spellCheck="false"
  />
</div>
`;
